### PR TITLE
Require the secret config to be configured

### PR DESCRIPTION
**Pull request by CarlSchwan**
If it's not configured the instance will look like it is working but
various features will silently break (end to end encryption, setting
alternate email and probably more).

One issue is that changing the secret from empty to something will
break various other stuff (app token). I don't think there is a good way
to solve this issue other than breaking early instead of having to
handle a painful migration later on.

image

### CODE REVIEW

**Review by szaimen — APPROVED** [review]

**Comment by CarlSchwan** [comment]
This breaks unit tests :( :( :(

**Comment by szaimen** [comment]
Still breaking many tests?

File: lib/private/legacy/OC_Util.php [hunk 1]
```diff
@@ -969,6 +969,13 @@ public static function checkServer(\OC\SystemConfig $config) {
 			];
 		}
 
+		if ($config->getValue('secret', '') === '') {
+			$errors[] = [
+				'error' => $l->t('The required \'secret\' config variable is not configued in the config.php file'),
```
Comment (come-nc) [inline lib/private/legacy/OC_Util.php:974]:
```suggestion
				'error' => $l->t('The required \'secret\' config variable is not configued in the config.php file.'),
```
Based on the screenshot it feels a dot is missing there given there is one on the other line.

Maybe we should have some requiredConfigVars array that lists the required vars? Is secret really the only one? Feels unlikely.

**Comment by juliushaertl** [comment]
Generally nice to have this checked 👍 We just should probably not backport and only show a setup warning on stable releases so that admins already get aware that this might be an issue.

The one failing test case looks interesting, I'm wondering a bit why the occ installation does not add the secret then:
Screenshot 2022-03-08 at 16 13 38

**Comment by szaimen** [comment]
> only show a setup warning on stable releases so that admins already get aware that this might be an issue.

shall I create a setup check for this?

**Comment by szaimen** [comment]
> > only show a setup warning on stable releases so that admins already get aware that this might be an issue.
> 
> shall I create a setup check for this?

I mean a setup check could be an alternative approach to this instead of breaking hard...

File: tests/travis/install.sh [hunk 2]
```diff
@@ -44,6 +44,7 @@ echo "Using database $DATABASENAME"
 cat > ./tests/autoconfig-sqlite.php <<DELIM
 <?php
 \$AUTOCONFIG = array (
+  'secret' => 'actually-not-secret',
```
Comment (juliushaertl) [inline tests/travis/install.sh:47]: Can we detect that and generate one during installation? It seems we also advertise the autoconfig in our docs https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/automatic_configuration.html#automatic-setup

Comment (CarlSchwan) [reply tests/travis/install.sh]: I removed this code for the setup and it's still working. Looking closer to the SetupController, even when using the autoconfig feature it still creates a secret and passwordhash

**Review by come-nc — APPROVED** [review]

**Review by szaimen — APPROVED** [review]

**Comment by szaimen** [comment]
> > only show a setup warning on stable releases so that admins already get aware that this might be an issue.
> 
> shall I create a setup check for this?

ping

**Comment by blizzz** [comment]
state is still set to "developing"?

**Comment by CarlSchwan** [comment]
> > > only show a setup warning on stable releases so that admins already get aware that this might be an issue.
> > 
> > 
> > shall I create a setup check for this?
> 
> ping

The idea is to have a migration path, see https://github.com/nextcloud/server/pull/31499

**Comment by szaimen** [comment]
So merge then?

**Pull request merged**